crates/red_knot/src/semantic/types/infer.rs

```diff
@@ -68,6 +89,37 @@ where
     }
 }
 
+/// Infer type for a ConstrainedDefinition (intersection of the definition type and the
+/// constraints)
+#[tracing::instrument(level = "trace", skip(db))]
+pub fn infer_constrained_definition_type(
+    db: &dyn SemanticDb,
+    symbol: GlobalSymbolId,
+    constrained_definition: ConstrainedDefinition,
+) -> QueryResult<Type> {
+    let ConstrainedDefinition {
+        definition,
+        constraints,
+    } = constrained_definition;
+    let index = semantic_index(db, symbol.file_id)?;
+    let parsed = parse(db.upcast(), symbol.file_id)?;
+    let mut intersected_types = vec![infer_definition_type(db, symbol, definition)?];
+    for constraint in constraints {
+        if let Some(constraint_type) = infer_constraint_type(
+            db,
+            symbol,
+            index.resolve_expression_id(parsed.syntax(), constraint),
+        )? {
+            intersected_types.push(constraint_type);
+        }
+    }
+    let jar: &SemanticJar = db.jar()?;
+    Ok(jar
+        .type_store
+        .add_intersection(symbol.file_id, &intersected_types, &[]))
+}
+
```

Comment:
Hmm, that becomes interesting in a salsa world where all types must be inferred upfront for each scope. What I have today is that each scope stores the types of each definition. 

I guess hwo this would work is that the intersection type would be defined in the scope that uses the definition (where the narrowing is happening) and not in the scope where the definition is defined. However, that can add a dependency from the usage scope to the AST nodes of the inference constraints, which might be undesired. Because I assume the expressions can come from many different scopes. But maybe that's just how it is.

Reply:
All of the CFG logic is internal to a scope, it doesn't cross scopes. There is no such thing as finding a definition in a different scope for a name. Currently this is kind of implicit in how we build the CFG; we start over with the Start node every time we enter a new scope, so there is no possible way to reach a different scope when traversing the CFG. All scopes share the same Start node, though this is really just an implementation optimization since the Start node has no data. We could possibly make this more explicit in having a separate FlowGraph structure for each scope instead.

I don't have support implemented yet for free/nonlocal/global variables. This first depends on doing the symbol table analysis so that we know when we are resolving a free/nonlocal/global variable. How this should work depends on how much narrowing we want to do on such variables, which is a correctness vs usability tradeoff -- users may expect narrowing of these variables, even where it's really not sound (because some intervening call could have changed the value in the other scope.) But assuming we want to narrow them, one way it could work is that we insert a synthetic "definition" at the start of the local scope for the free/nonlocal/global variable, and that definition would resolve to the public type of the symbol in the scope where it's defined.

I don't think we will ever want the CFG traversal itself to cross scopes.